src/Cryptol/Backend/Monad.hs

```diff
@@ -92,7 +113,13 @@ combineCallStacks ::
   CallStack {- ^ call stack of the application context -} ->
   CallStack {- ^ call stack of the function being applied -} ->
   CallStack
-combineCallStacks appstk fnstk = appstk <> dropCommonPrefix appstk fnstk
+combineCallStacks appstk fnstk = CombineCallStacks appstk fnstk
```

Comment:
I'm not sure how common this is, but is it worth checking for the special cases where one of the arguments is the `EmptyCallStack`?

Reply:
Good idea. That would cost very little to check for, and it would strictly reduce the memory consumption whenever that rule matches.

Reply:
I added special cases for `EmptyCallStack`. Looking at `+RTS -s`, any performance differences (on my use case at least) are lost in the noise. But we might as well keep it in; it only reduces total allocation by 0.01% but it might be more on other examples.